lib/THC/THCTensorRandom.cu

```diff
@@ -359,4 +359,405 @@ THC_API void THCudaTensor_cauchy(THCState* state, THCudaTensor *self_, double me
+                      int categories,
+                      float* dist) {
+  extern __shared__ float smem[];
+
+  for (long curDist = blockIdx.x;
+       curDist < distributions; curDist += gridDim.x) {
+    // Each block handles one distribution
+    // First pass, find the total sum of the distribution
+    float sum = 0.0f;
+    for (int cat = threadIdx.x; cat < categories; cat += blockDim.x) {
+      sum += dist[curDist * categories + cat];
+    }
+
+    // threadIdx.x == 0 has the sum value from this
+    sum = reduceBlock(smem, blockDim.x, sum, thrust::plus<float>(), 0.0f);
+
+    // Broadcast sum and sample value
+    if (threadIdx.x == 0) {
+      smem[0] = sum;
+      smem[1] = dest[curDist];
+    }
+    __syncthreads();
+
+    sum = smem[0];
+    float sample = smem[1];
+    __syncthreads();
+
+    if (sum == 0.0f || sample == 0.0f) {
+      // Choose the first element
+      if (threadIdx.x == 0) {
+        dest[curDist] = 1;
+      }
+
+      continue;
+    }
+
+    int chunks = THCCeilDiv(categories, (int) blockDim.x);
+    float prevHighProb = 0.0f;
+
+    for (int chunk = 0; chunk < chunks; ++chunk) {
+      // All threads in bounds load a value
+      int cat = chunk * blockDim.x + threadIdx.x;
+
+      float val =
+        cat < categories ? dist[curDist * categories + cat] / sum : 0.0f;
+      smem[threadIdx.x] = val;
+      __syncthreads();
+
+      // Perform an inclusive prefix sum of the shared memory contents
+      for (int offset = 1; offset < blockDim.x; offset *= 2) {
+        float val = 0.0f;
+
+        if (threadIdx.x >= offset) {
+          val = smem[threadIdx.x - offset] + smem[threadIdx.x];
+        }
+
+        __syncthreads();
+        if (threadIdx.x >= offset) {
+          smem[threadIdx.x] = val;
+        }
+        __syncthreads();
+      }
+
+      // Each thread will check to see if the sample falls in its
+      // bucket
+      float curBucket =
+        smem[threadIdx.x] + prevHighProb;
+      float prevBucket =
+        threadIdx.x == 0 ? prevHighProb : smem[threadIdx.x - 1] + prevHighProb;
+      bool inBucket =
+        (cat < categories) && (sample <= curBucket) && (sample > prevBucket);
+
+      if (inBucket) {
+        // We're done; we have the sample
+        // Torch indices are 1-based
+        // FIXME: broadcast exit flag?
+        dest[curDist] = cat + 1;
+      }
+
+      // Store the previous scan's high value for future use
+      prevHighProb += smem[blockDim.x - 1];
+
+      __syncthreads();
+    }
+  }
+}
+
+__global__ void
+sampleMultinomialWithReplacement(curandStateMtgp32* state,
+                                 int totalSamples,
+                                 float* dest,
+                                 long distributions,
+                                 int categories,
+                                 float* normDistPrefixSum) {
+  // Binary search is by its nature warp divergent, so all threads
+  // in a warp compute the save value. It's important that all threads
```

Comment:
What do you mean by "all threads in a warp compute the save value"? Only one thread in each warp actually computes a value, right?

Have you actually tried using more than one thread per warp? The depth of your binary search is always the same, isn't it? So the only divergence is in the `midVal < val` condition, which you could re-write using the ternary operator. Would be interesting to see if that buys you anything.